File: Data Structures & Algorithms/duplicate-integer/submission-4.py
# class Solution:
#     def hasDuplicate(self, nums: List[int]) -> bool:
#         seen = set()
#         for num in nums:
#             if num in seen:
#                 return True
#             seen.add(num)
#         return False

class Solution:
    def hasDuplicate(self, nums: List[int]) -> bool:
        nums.sort()
        for n in range(1, len(nums)):
            if nums[n] == nums[n - 1]:
                return True
        return False
            

File: Data Structures & Algorithms/products-of-array-discluding-self/submission-4.js
class Solution {
    /**
     * @param {number[]} nums
     * @return {number[]}
     */
    // productExceptSelf(nums) {
    //     const totalProduct = nums.reduce((acc, curr) => acc * curr, 1);

    //     return nums.map((num) => totalProduct / num);
    // }

    productExceptSelf(nums) {
        const preff = new Array(nums.length)
        const suff = new Array(nums.length)
        const res = new Array(nums.length)

        preff[0] = 1
        suff[nums.length -1] = 1

        for(let i =1; i < nums.length; i++){
            preff[i] = preff[i-1] * nums[i-1]
        }

        for(let i = nums.length -2; i>=0; i--){
            suff[i] = suff[i +1]* nums[i+1]
        }

        for(let i=0; i<nums.length; i++){
            res[i] = preff[i] *suff[i]
        }

        return res

    }

}
